### ISSUE TITLE
Kafka sources connecting to the same broker should shared

### ISSUE THREAD

**Issue by guswynn** [MaterializeInc/materialize#26798]
Related to: https://github.com/MaterializeInc/cloud/issues/9260

Thread with context: https://materializeinc.slack.com/archives/C01CFKM1QRF/p1713984529700059?thread_ts=1713921112.345979&cid=C01CFKM1QRF

Just like ssh tunnels, kafka connections should be shared between all sources (and worker threads of those sources) that use them. This means we need a global registry of active kafka clients that sources can share from

Some caveats to figure out are:
- How to distribute source statuses and ensure all sources are marked in `mz_source_status_history` correctly
- How to create independent consumers from shared clients
- Figuring out how `rdkafka` managers connections across many consumers that share clients.

One change that has a similar implementation, but is less risky is only sharing clients between the workers of a source. This and true-global sharing could be configurable with an LD parameter.

**Comment by benesch** [comment]
> One change that has a similar implementation, but is less risky is only sharing clients between the workers of a source.

After talking with @guswynn live just now, strong agree that we should start with this if it's easy. Should make a meaningful dent in the number of sockets and connections that we create.

I think every thread right now also has its own metadata fetching consumer/thread? If that's still the case, it's worth optimizing too.

**Comment by guswynn** [comment]
Yeah, definitely; I originally though that sharing a `rdkafka::Client` was sufficient, but you actually need to share a `Consumer`. This is extremely difficult cross-source, but easier to do cross-worker within a source
